utils/ProfileProvider.tsx

```diff
@@ -0,0 +1,99 @@
+'use client';
+
+import React, {
+  createContext,
+  ReactNode,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+import { upsertProfile } from '@/api/supabase/queries/profiles';
+import { Profile } from '@/types/schema';
+
+// Define a placeholder user ID for development purposes
+const placeholderUserId = '2abd7296-374a-42d1-bb4f-b813da1615ae';
+
+interface ProfileContextType {
+  profileData: Profile | null;
+  profileReady: boolean;
+  setProfile: (newProfileData: Partial<Profile>) => Promise<void>;
+  loadProfile: () => Promise<void>;
+}
+
+const ProfileContext = createContext<ProfileContextType | undefined>(undefined);
+
+export const useProfile = () => {
+  const context = useContext(ProfileContext);
+  if (!context) {
+    throw new Error('useProfile must be used within a ProfileProvider');
+  }
+  return context;
+};
+
+export const ProfileProvider: React.FC<{ children: ReactNode }> = ({
+  children,
+}) => {
+  const [profileData, setProfileData] = useState<Profile | null>(null);
+  const [profileReady, setProfileReady] = useState<boolean>(false);
+
+  const loadProfile = useCallback(async () => {
+    setProfileReady(false);
+
+    try {
+      const profile: Profile = {
+        user_id: placeholderUserId,
+        state: '',
+        user_type: '',
+        has_plot: false,
+      };
+      // Fetch or upsert the profile for the placeholder user ID
+      const fetchedProfile = await upsertProfile(profile);
+      setProfileData(fetchedProfile);
+    } catch (error) {
+      console.error('Error loading profile:', error);
+    } finally {
+      setProfileReady(true);
+    }
+  }, []);
+
+  const setProfile = useCallback(
+    async (newProfileData: Partial<Profile>) => {
```

Comment:
i think we should be taking in a Profile rather than Partial<Profile>. Do you anticipate using this function anywhere besides the end of the onboarding flow? We want to require the user to fill out all onboarding questions 

If we really wanted an updateProfile function, we can probably create a separate one. But for setProfile, i think all fields of Profile should be filled out, and it should error otherwise.  
@ronniebeggs @pragyakallanagoudar thoughts?